File: src/y2025/day08.rs
//! <https://adventofcode.com/2025/day/8>
use {
    crate::{
        framework::{AdventOfCode, ParseError, aoc},
        geometric::Dim3,
    },
    std::collections::HashMap,
};

#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Puzzle {
    line: Vec<Dim3<usize>>,
    dists: Vec<(usize, (usize, usize))>,
}

mod parser {
    use {
        crate::{geometric::Dim3, parser::parse_usize},
        winnow::{ModalResult, Parser, ascii::newline, combinator::separated},
    };

    fn parse_line(s: &mut &str) -> ModalResult<Dim3<usize>> {
        separated(1.., parse_usize, ",")
            .map(|v: Vec<usize>| (v[0], v[1], v[2]))
            .parse_next(s)
    }
    pub fn parse(s: &mut &str) -> ModalResult<Vec<Dim3<usize>>> {
        separated(1.., parse_line, newline).parse_next(s)
    }
}

#[aoc(2025, 8)]
impl AdventOfCode for Puzzle {
    fn prepare(&mut self, mut input: &str) -> Result<(), ParseError> {
        self.line = parser::parse(&mut input)?;
        let num_nodes = self.line.len();
        let mut occs: Vec<usize> = vec![usize::MAX; num_nodes];
        let mut dists: Vec<Vec<(usize, (usize, usize))>> = Vec::new();
        for i in 0..num_nodes {
            let mut min = usize::MAX;
            let mut min_index = 0;
            let mut ret: Vec<(usize, (usize, usize))> = Vec::new();
            for j in i + 1..num_nodes {
                let x = self.line[i];
                let y = self.line[j];
                let d = [
                    x.0.abs_diff(y.0).pow(2),
                    x.1.abs_diff(y.1).pow(2),
                    x.2.abs_diff(y.2).pow(2),
                ]
                .iter()
                .sum::<usize>();
                if d < min {
                    min = d;
                    min_index = j;
                }
                ret.push((d, (i, j)));
            }
            occs[i] = occs[i].min(min);
            occs[min_index] = occs[min_index].min(min);
            dists.push(ret);
        }
        let thr = occs.into_iter().max().unwrap();
        let mut d = dists
            .into_iter()
            .flat_map(|l| l.into_iter().filter(|r| r.0 <= thr).collect::<Vec<_>>())
            .collect::<Vec<_>>();
        d.sort_unstable();
        self.dists = d;
        Ok(())
    }
    fn part1(&mut self) -> Self::Output1 {
        let limit = self
            .get_config()
            .alt
            .as_ref()
            .map_or(1000_usize, |_| 10_usize);
        let mut group_heap: Vec<usize> = vec![0];
        let mut membership: Vec<usize> = vec![0; self.line.len()];
        let mut new_group: usize = 0;
        for (_, (i, j)) in self.dists.iter().take(limit) {
            let mut g1 = membership[*i];
            while group_heap[g1] != 0 {
                g1 = group_heap[g1];
            }
            let mut g2 = membership[*j];
            while group_heap[g2] != 0 {
                g2 = group_heap[g2];
            }
            match (g1 == 0, g2 == 0) {
                (false, false) => {
                    if g1 != g2 {
                        let a = g1.min(g2);
                        let b = g1.max(g2);
                        group_heap[b] = a;
                    }
                }
                (false, true) => {
                    membership[*j] = membership[*i];
                }
                (true, false) => {
                    membership[*i] = membership[*j];
                }
                (true, true) => {
                    new_group += 1;
                    group_heap.push(0);
                    // assert_eq!(new_group, group_heap.len() - 1);
                    membership[*i] = new_group;
                    membership[*j] = new_group;
                }
            }
        }
        let mut groups: HashMap<usize, Vec<usize>> = HashMap::new();
        for (i, rg) in membership.iter().enumerate() {
            let mut g = *rg;
            while group_heap[g] != 0 {
                g = group_heap[g];
            }
            groups.entry(g).or_default().push(i);
        }
        let mut gv = groups
            .iter()
            .map(|(id, l)| if *id == 0 { 1 } else { l.len() })
            .collect::<Vec<_>>();
        gv.sort();
        gv[gv.len() - 3..].iter().product()
    }
    fn part2(&mut self) -> Self::Output2 {
        let mut group_heap: Vec<usize> = vec![0];
        let mut membership: Vec<usize> = vec![0; self.line.len()];
        let mut new_group: usize = 0;
        let mut num_groups = 0;
        for (_, (i, j)) in self.dists.iter() {
            let mut g1 = membership[*i];
            while group_heap[g1] != 0 {
                g1 = group_heap[g1];
            }
            let mut g2 = membership[*j];
            while group_heap[g2] != 0 {
                g2 = group_heap[g2];
            }
            match (g1 == 0, g2 == 0) {
                (false, false) => {
                    if g1 != g2 {
                        let a = g1.min(g2);
                        let b = g1.max(g2);
                        group_heap[b] = a;
                        num_groups -= 1;
                    }
                }
                (false, true) => {
                    membership[*j] = membership[*i];
                }
                (true, false) => {
                    membership[*i] = membership[*j];
                }
                (true, true) => {
                    new_group += 1;
                    num_groups += 1;
                    group_heap.push(0);
                    membership[*i] = new_group;
                    membership[*j] = new_group;
                }
            }
            if num_groups == 1 && membership.iter().filter(|g| **g != 0).count() == self.line.len()
            {
                return self.line[*i].0 * self.line[*j].0;
            }
        }
        unreachable!()
    }
}

File: Swift/Sources/Y2025/day08.swift
import Collections
import Parsing
import Utils

private struct Ordered<T>: Comparable {
    let priority: Int
    let value: T
    static func < (lhs: Ordered<T>, rhs: Ordered<T>) -> Bool {
        lhs.priority < rhs.priority
    }
    static func == (lhs: Ordered<T>, rhs: Ordered<T>) -> Bool {
        lhs.priority == rhs.priority
    }
}

private func part1(_ input: [Pos3]) -> Int {
    let limit = input.count == 20 ? 10 : 1000
    var heap: Heap<Ordered<(Int, Int)>> = Heap()
    for i in input.indices {
        for j in input.indices {
            if j <= i { continue }
            let p = (input[i] - input[j]).abs()
            let dist = p.z * p.z + p.y * p.y + p.x * p.x
            if heap.count >= limit {
                if heap.max!.priority > dist {
                    heap.removeMax()
                    heap.insert(Ordered(priority: dist, value: (i, j)))
                }
            } else {
                heap.insert(Ordered(priority: dist, value: (i, j)))
            }
        }
    }
    assert(heap.count == limit)
    var groupHeap: [Int] = [0]
    var membership: [Int] = Array(repeating: 0, count: input.count)
    var newGroup = 0
    while let d = heap.popMin() {
        let (i, j) = d.value
        var g1 = membership[i]
        while groupHeap[g1] != 0 {
            g1 = groupHeap[g1]
        }
        var g2 = membership[j]
        while groupHeap[g2] != 0 {
            g2 = groupHeap[g2]
        }
        switch (g1 == 0, g2 == 0) {
        case (false, false):
            if g1 != g2 {
                groupHeap[max(g1, g2)] = min(g1, g2)
            }
        case (false, true):
            membership[j] = membership[i]
        case (true, false):
            membership[i] = membership[j]
        case (true, true):
            newGroup += 1
            groupHeap.append(0)
            membership[i] = newGroup
            membership[j] = newGroup
        }
    }
    var groups: [Int: [Int]] = [:]
    for i in membership.indices {
        var g = membership[i]
        while groupHeap[g] != 0 {
            g = groupHeap[g]
        }
        groups[g, default: []].append(i)
    }
    var gv = groups.map { (id, l) in if id == 0 { 1 } else { l.count } }
    gv.sort(by: { $0 > $1 })
    return gv[0] * gv[1] * gv[2]
}

private func part2(_ input: [Pos3]) -> Int {
    let n = input.count
    var occurred: [Int] = Array(repeating: Int.max, count: input.count)
    var dists: [[Ordered<(Int, Int)>]] = []
    for i in input.indices {
        var minDist = Int.max
        var minIndex = 0
        var ret: [Ordered<(Int, Int)>] = []
        for j in input.indices {
            if j <= i { continue }
            let p = (input[i] - input[j]).abs()
            let dist = p.z * p.z + p.y * p.y + p.x * p.x
            if dist < minDist {
                minDist = dist
                minIndex = j
            }
            ret.append(Ordered(priority: dist, value: (i, j)))
        }
        occurred[i] = min(occurred[i], minDist)
        occurred[minIndex] = min(occurred[minIndex], minDist)
        dists.append(ret)
    }
    let thr = occurred.max()!
    var pickedDists = dists.flatMap { l in
        l.filter { $0.priority <= thr }
    }
    pickedDists.sort(by: { $0.priority < $1.priority })
    var groupHeap: [Int] = [0]
    var membership: [Int] = Array(repeating: 0, count: input.count)
    var newGroup = 0
    var numGroups = 0
    for d in pickedDists {
        let (i, j) = d.value
        var g1 = membership[i]
        while groupHeap[g1] != 0 {
            g1 = groupHeap[g1]
        }
        var g2 = membership[j]
        while groupHeap[g2] != 0 {
            g2 = groupHeap[g2]
        }
        switch (g1 == 0, g2 == 0) {
        case (false, false):
            if g1 != g2 {
                groupHeap[max(g1, g2)] = min(g1, g2)
                numGroups -= 1
            }
        case (false, true):
            membership[j] = membership[i]
        case (true, false):
            membership[i] = membership[j]
        case (true, true):
            newGroup += 1
            numGroups += 1
            groupHeap.append(0)
            membership[i] = newGroup
            membership[j] = newGroup
        }
        if numGroups == 1 && membership.filter({ $0 != 0 }).count == n {
            return input[i].z * input[j].z
        }
    }
    fatalError()
}

public func day08(_ data: String) throws {
    let line_parser: some Parser<Substring, Pos3> = Parse {
        Int.parser()
        ","
        Int.parser()
        ","
        Int.parser()
    }.map { Pos3(z: $0, y: $1, x: $2) }
    let parser: some Parser<Substring, [Pos3]> = Many {
        line_parser
    } separator: {
        "\n"
    } terminator: {
        "\n"
    }
    let input = try parser.parse(data)
    let sum1 = part1(input)
    let sum2 = part2(input)
    print("Part 1: \(sum1)")
    print("Part 2: \(sum2)")
}
